substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+
+			// Ensure delegator is sane.
+			ensure!(!Self::is_agent(&delegator), Error::<T>::NotAllowed);
+			ensure!(!Self::is_delegator(&delegator), Error::<T>::NotAllowed);
+			ensure!(Self::not_direct_staker(&delegator), Error::<T>::AlreadyStaking);
+
+			// ensure agent is sane.
+			ensure!(Self::is_agent(&agent), Error::<T>::NotAgent);
+
+			// and has enough delegated balance to migrate.
+			let proxy_delegator = Self::sub_account(AccountType::ProxyDelegator, agent);
+			let balance_remaining = Self::held_balance_of(&proxy_delegator);
+			ensure!(balance_remaining >= amount, Error::<T>::NotEnoughFunds);
+
+			Self::do_migrate_delegation(&proxy_delegator, &delegator, amount)
+		}
+
+		/// Delegate given `amount` of tokens to an `Agent` account.
+		///
+		/// If `origin` is the first time delegator, we add them to state. If they are already
+		/// delegating, we increase the delegation.
+		///
+		/// Conditions:
+		/// - Delegators cannot delegate to more than one agent.
+		/// - The `agent` account should already be registered as such. See [`Self::register_agent`]
+		pub fn delegate_to_agent(
+			origin: OriginFor<T>,
+			agent: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let delegator = ensure_signed(origin)?;
+
+			// ensure delegator is sane.
+			ensure!(
+				Delegation::<T>::can_delegate(&delegator, &agent),
+				Error::<T>::InvalidDelegation
+			);
+			ensure!(Self::not_direct_staker(&delegator), Error::<T>::AlreadyStaking);
+
+			// ensure agent is sane.
+			ensure!(Self::is_agent(&agent), Error::<T>::NotAgent);
+
+			// add to delegation
+			Self::do_delegate(&delegator, &agent, amount)?;
+
+			// bond the newly delegated amount to `CoreStaking`.
+			Self::do_bond(&agent, amount)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		#[cfg(feature = "try-runtime")]
+		fn try_state(_n: BlockNumberFor<T>) -> Result<(), sp_runtime::TryRuntimeError> {
+			Self::do_try_state()
+		}
+
+		fn integrity_test() {}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Derive a (keyless) pot account from the given agent account and account type.
+	pub(crate) fn sub_account(account_type: AccountType, agent: T::AccountId) -> T::AccountId {
+		T::PalletId::get().into_sub_account_truncating((account_type, agent.clone()))
+	}
+
+	/// Balance of a delegator that is delegated.
+	pub(crate) fn held_balance_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Currency::balance_on_hold(&HoldReason::Delegating.into(), who)
+	}
+
+	/// Returns true if who is registered as an `Agent`.
+	fn is_agent(who: &T::AccountId) -> bool {
+		<Agents<T>>::contains_key(who)
+	}
+
+	/// Returns true if who is delegating to an `Agent` account.
+	fn is_delegator(who: &T::AccountId) -> bool {
+		<Delegators<T>>::contains_key(who)
+	}
+
+	/// Returns true if who is not already staking on [`Config::CoreStaking`].
+	fn not_direct_staker(who: &T::AccountId) -> bool {
+		T::CoreStaking::status(who).is_err()
+	}
```

Comment:
As said above, I think it's simpler to avoid negations in the method names. e.g. !not_direct_staker becomes double negation and it's harder to read.

```suggestion
	/// Returns true if who is already staking on [`Config::CoreStaking`].
	fn is_direct_staker(who: &T::AccountId) -> bool {
		T::CoreStaking::status(who).is_ok()
	}
```